### PR TITLE
The bootstrap now really works

### PR DESCRIPTION
**Pull request by samoht**
Before, only the `assemblage` library was built (wrongly). Now all the libraries and the binaries are successfully built. The tests result are not very meaningful yet as they require #161 to work first.

### CODE REVIEW

File: lib/as_action_ocaml.ml [hunk 1]
```diff
@@ -67,7 +67,7 @@ let compute_deps_ml = compute_deps `Ml
 
 let compile_mli
     ?(needs = []) ?(pkgs = nop) ?(args = []) ~ocamlc ~annot ~incs ~src () =
-  let ctx = As_ctx.v [`OCaml; `Compile; `Src `Mli] in
+  let ctx = As_ctx.v [`OCaml; `Compile; `Src `Mli; `Target `Byte] in
```
Comment (dbuenzli) [inline lib/as_action_ocaml.ml]: Why did you add this ? `mli` may be compiled when working on native code generation. In my opinion either `mli` compilation should remain `` `Target`` agnostic or we should distinguish the two (which one could argue should be done because e.g. extension nodes may do something different in the two target, but I don't think it's really feasible with the way the toolchain works).

File: driver/builder_makefile.ml [hunk 2]
```diff
@@ -192,12 +204,48 @@ let mk_gen_prepare gen =
...
+  generator proj
+  |> keys
+  |> mk_phony
+  |> mk_all
+  |> mk_test
+  |> (fun gen -> List.fold_left mk_part gen (Project.parts gen.proj))
+  |> mk_dirs
+  |> mk_ds
+  |> mk_prepares
+  |> ret
```
Comment (dbuenzli) [inline driver/builder_makefile.ml]: Personally I don't find that kind of code more readable than the previous style. Point-free style has sometimes its uses in DSLs but in general it puts a larger cognitive burden on me (and has the effect of introducing needless functions in that case, and why name `ret` and not the other ?).  I prefer to read OCaml functions as sequences of `let` bindings where variables are explicitely threaded.

File: lib/as_action_ocaml.ml (hunk 1 above)
Comment (dbuenzli) [inline lib/as_action_ocaml.ml]: Note also that the `ocamlc` argument will be `ocamlopt` if the byte code compiler is not available see here https://github.com/samoht/assemblage/blob/master/lib/as_part_unit.ml#L128-L129

File: driver/builder_makefile.ml (hunk 2 above)
Comment (samoht) [inline driver/builder_makefile.ml]: ok I'll revert that.

File: lib/as_action_ocaml.ml (hunk 1 above)
Comment (samoht) [inline lib/as_action_ocaml.ml]: Because this was causing missing includes dirs for external packages. But yes, that was probably not the best place to fix it, might be better to change the caller's `incs`instead.

Comment (dbuenzli) [inline lib/as_action_ocaml.ml]: The right place to solve the problem is [here](https://github.com/samoht/assemblage/blob/master/lib/as_ocamlfind.ml#L75-L81) I think.

Comment (dbuenzli) [inline lib/as_action_ocaml.ml]: You should add a case for `` `OCaml; `Compile; `Src `Mli;`` with the right package information.

Comment (samoht) [inline lib/as_action_ocaml.ml]: but it's not clear which ocamlfind command to run then ... byte or native? I guess in practice it doesn't really make a difference...

Comment (dbuenzli) [inline lib/as_action_ocaml.ml]: Gah, I forgot how all this system is completely annoying...

Comment (dbuenzli) [inline lib/as_action_ocaml.ml]: Maybe the most future proof of solving that problem is to add a `target` argument to this function with either `` `Target `Byte`` or `` `Target `Native` that you add to the context according to which `ocamlc` is passed as an argument (i.e. define that [here](https://github.com/samoht/assemblage/blob/master/lib/as_part_unit.ml#L128-L129)).

Comment (dbuenzli) [inline lib/as_action_ocaml.ml]: Yes I think that's the right solution in fact we are already doing that kind of distinction for project [local paths](https://github.com/samoht/assemblage/blob/master/lib/as_part_unit.ml#L147). It may not be meaningful now but it may become in the future. End user that want to address the computation should use the context `` `[`OCaml; `Compile; `Src `Mli; ]`` (without the target specification) for now (this will work since contexts are matched by subsetting) as there is concretly no distinction between the two targets. The one that is selected is based purely on the availability of tools at hand.

Comment (samoht) [inline lib/as_action_ocaml.ml]: I've just rebased the patch set, and this is indeed much simpler now.

File: lib/as_ocamlfind.ml [hunk 3]
```diff
@@ -72,16 +74,20 @@ let parse_lines (byte, native, pp) =
     native_incs; native_objs; native_link;
     pp_incs; pp_objs; }
 
-let pkg_lookups ocamlfind name =
+let pkg_lookups ocamlfind native name =
```
Comment (dbuenzli) [inline lib/as_ocamlfind.ml]: @samoht that's not what I was suggesting. I don't like the fact the `native` logic crept in `ocamlfind.ml` this is obfuscating what is going on and should be explicitely handled at the `unit` level where all the `native` conditioning is already done.

Comment (samoht) [inline lib/as_ocamlfind.ml]: Hum yes, I think I see what you mean. I'll try it.

Comment (dbuenzli) [inline lib/as_ocamlfind.ml]: Basically this [line](https://github.com/samoht/assemblage/blob/master/lib/as_part_unit.ml#L128-L129) should tuple the compiler and the corresponding `` `Target``. The latter must be added to the context [here](https://github.com/samoht/assemblage/pull/163#discussion-diff-24478042R70) and there is nothing to touch in `ocamlfind.ml`.

**Comment by samoht** [comment]
ok, pushed a new version. Still not totally sure it is what you expect, but I'm not sure to see how to do differently...

File: lib/as_ocamlfind.ml [hunk 4]
```diff
@@ -77,6 +79,8 @@ let pkg_lookups ocamlfind name =
   [ As_ctx.v [`OCaml; `Pp], p.pp_incs @ p.pp_objs;
     As_ctx.v [`OCaml; `Compile; `Target `Byte], p.byte_incs;
     As_ctx.v [`OCaml; `Compile; `Target `Native], p.native_incs;
+    As_ctx.v [`OCaml; `Compile; `Target `Byte; `Src `Mli], p.byte_incs;
+    As_ctx.v [`OCaml; `Compile; `Target `Native; `Src `Mli], p.native_incs;
```
Comment (dbuenzli) [inline lib/as_ocamlfind.ml]: Yes maybe it even better like that.
